### PR TITLE
v4.1.x: pml/ucx: fix datatype refcount

### LINKED ISSUE #9138: 4.x: Segfault in mca_pml_ucx_start if MPI_Type_Free before MPI_Start
## Background information

I'm building OpenMPI from source (https://download.open-mpi.org/release/open-mpi/v4.1/openmpi-4.1.1.tar.bz2):

```
configure --enable-orterun-prefix-by-default --enable-mpirun-prefix-by-default --without-verbs --without-cs-fs --enable-mca-no-build=btl-uct --prefix=/src/openmpi/install --enable-debug
```

OpenUCX was also built from source with
```
configure --prefix=/src/openmpi/install-ucx --enable-debug --enable-cma --with-verbs --with-rdmacm=/usr/local --enable-mt
```

Both OpenUCX 1.10.1 and 1.11.x (84dcd8033) allow to reproduce the bug.

* Operating system/version: RHEL7
* Network type: mlx4

-----------------------------

## Details of the problem

I get a segfault in `MPI_Start`:

```
#0  0x0000000000000000 in ?? ()
#1  0x00007fffef96f475 in ucp_dt_recv_state_init (dt_count=1, dt=21739655, buffer=0x0, dt_state=0x13afb40) at /src/openmpi/ucx/src/ucp/dt/dt.inl:147
#2  ucp_tag_recv_common (debug_name=0x7fffef9da686 'recv_nbx\000', param=0x7ffffffed740, rdesc=0x0, req=0x13afb00, tag_mask=18446744073709551615, tag=76965813946417156, 
    datatype=21739655, count=1, buffer=0x0, worker=0x131c2a0) at /src/openmpi/ucx/src/ucp/tag/tag_recv.c:88
#3  ucp_tag_recv_nbx (worker=0x131c2a0, buffer=0x0, count=1, tag=76965813946417156, tag_mask=18446744073709551615, param=0x7ffffffed740)
    at /src/openmpi/ucx/src/ucp/tag/tag_recv.c:224
#4  0x00007fffef96db9d in ucp_tag_recv_nb (worker=0x131c2a0, buffer=0x0, count=1, datatype=21739655, tag=76965813946417156, tag_mask=18446744073709551615, 
    cb=0x7ffff4060b4c <mca_pml_ucx_precv_completion>) at /src/openmpi/ucx/src/ucp/tag/tag_recv.c:198
#5  0x00007ffff405f8d8 in mca_pml_ucx_start (count=1, requests=0x7ffffffed878) at /src/openmpi/openmpi-4.1.1/ompi/mca/pml/ucx/pml_ucx.c:1123
#6  0x00007ffff7cc62b2 in PMPI_Start (request=0x7ffffffed878) at pstart.c:76
#7  0x00007ffff7f8e139 in ompi_start_f (request=0x14a3f10, ierr=0x7fffffff5960) at pstart_f.c:75
#8  0x000000000043c571 in ...
#9  0x000000000045a0fb in main (argc=1, argv=0x7fffffffd929)
#10 0x00007ffff6744555 in __libc_start_main () from target:/lib64/libc.so.6
#11 0x00000000004037f9 in _start ()
```

So far I was not able to reproduce that on a trivial example that I could share with you. I hope you can guide me to find the root cause of that bug. My first guess is that `ucp_tag_recv_nb` arguments are wrong, that's why I chose to report the bug here and not at OpenUCX.

### PR DESCRIPTION
**Pull request by karasevb**
Fixes #9138

### CODE REVIEW

File: ompi/mca/pml/ucx/pml_ucx.c [hunk 1]
```diff
@@ -705,6 +706,7 @@ int mca_pml_ucx_isend_init(const void *buf, size_t count, ompi_datatype_t *datat
         OBJ_RETAIN(datatype);
     } else {
         req->datatype.datatype = mca_pml_ucx_get_datatype(datatype);
+        OMPI_DATATYPE_RETAIN(datatype);
```
Comment (yosefe) [inline ompi/mca/pml/ucx/pml_ucx.c:709]: do we need to release refcount somewhere?

**Review by yosefe — COMMENTED** [review]
why for 4.1 and not master?

File: ompi/mca/pml/ucx/pml_ucx.c (hunk 1 above)
Comment (karasevb) [reply ompi/mca/pml/ucx/pml_ucx.c:709]: Yes, will fix it.

**Comment by karasevb** [comment]
Closing this and opening the fixed new one against `master`.

**Pull request closed without merging**